File: lang/go/core/lws/management.go
package lws

import (
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/gorilla/websocket"
	lwsstate "github.com/local-web-services/local-web-services-go-core/lws/state"
)

var wsUpgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

// RegisterManagementAPI registers the management HTTP endpoints on the given mux.
func RegisterManagementAPI(mux *http.ServeMux, state *ServerState, shutdownCh chan struct{}) {
	// GET /_ldk/status
	mux.HandleFunc("/_ldk/status", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			http.NotFound(w, r)
			return
		}
		writeJSON(w, map[string]interface{}{"running": true, "providers": []string{}}, 200)
	})

	// POST /_ldk/reset
	mux.HandleFunc("/_ldk/reset", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.NotFound(w, r)
			return
		}
		state.Reset()
		state.ResetCapacity()
		writeJSON(w, map[string]string{"status": "ok"}, 200)
	})

	// GET /_ldk/logs
	mux.HandleFunc("/_ldk/logs", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			http.NotFound(w, r)
			return
		}
		logs := state.GetLogs()
		if logs == nil {
			logs = []LogEntry{}
		}
		writeJSON(w, map[string]interface{}{"logs": logs}, 200)
	})

	// GET /_ldk/ws/logs — WebSocket streaming log endpoint.
	// Sends existing log entries then polls for new ones every 50ms.
	mux.HandleFunc("/_ldk/ws/logs", func(w http.ResponseWriter, r *http.Request) {
		conn, err := wsUpgrader.Upgrade(w, r, nil)
		if err != nil {
			return
		}
		defer conn.Close() //nolint:errcheck

		sent := 0
		for {
			entries := state.GetLogs()
			for i := sent; i < len(entries); i++ {
				data, err := json.Marshal(wsLogEntry(entries[i]))
				if err != nil {
					continue
				}
				if err := conn.WriteMessage(websocket.TextMessage, data); err != nil {
					return
				}
			}
			sent = len(entries)
			time.Sleep(50 * time.Millisecond)
		}
	})

	// GET /_ldk/chaos and POST /_ldk/chaos
	mux.HandleFunc("/_ldk/chaos", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			status := state.GetAllChaosStatus()
			writeJSON(w, status, 200)
		case http.MethodPost:
			var body map[string]map[string]interface{}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			for service, config := range body {
				enabled, hasEnabled := config["enabled"]
				if hasEnabled {
					if b, ok := enabled.(bool); ok && !b {
						state.DisableChaos(service)
						continue
					}
				}

				nonEnabledKeys := 0
				for k := range config {
					if k != "enabled" {
						nonEnabledKeys++
					}
				}
				if nonEnabledKeys == 0 {
					state.EnableChaos(service)
					continue
				}

				rule := &ChaosRule{}
				if v, ok := config["error_rate"]; ok {
					rule.ErrorRate = toFloat64(v)
				}
				if v, ok := config["latency_min_ms"]; ok {
					rule.LatencyMinMs = int(toFloat64(v))
				}
				if v, ok := config["latency_max_ms"]; ok {
					rule.LatencyMaxMs = int(toFloat64(v))
				}
				if v, ok := config["error_code"]; ok {
					if s, ok := v.(string); ok {
						rule.ErrorCode = s
					}
				}
				state.EnableChaos(service)
				state.SetChaosRule(service, "*", rule)
			}
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// GET /_ldk/lifecycle and POST /_ldk/lifecycle
	mux.HandleFunc("/_ldk/lifecycle", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			// Return current lifecycle rules (not yet implemented for GET)
			writeJSON(w, map[string]interface{}{}, 200)
		case http.MethodPost:
			var body map[string]map[string]interface{}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			for svc, config := range body {
				rule := LifecycleRule{}
				if v, ok := config["enabled"].(bool); ok {
					rule.Enabled = v
				} else {
					rule.Enabled = true
				}
				if v, ok := config["create_dwell_ms"].(float64); ok {
					rule.CreateDwellMs = int(v)
				}
				if v, ok := config["delete_dwell_ms"].(float64); ok {
					rule.DeleteDwellMs = int(v)
				}
				state.SetLifecycleRule(svc, rule)
			}
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// GET /_ldk/capacity and POST /_ldk/capacity
	mux.HandleFunc("/_ldk/capacity", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			status := state.GetAllCapacityStatus()
			result := make(map[string]map[string]interface{})
			for svc, rule := range status {
				result[svc] = map[string]interface{}{"slots": rule.Slots}
			}
			writeJSON(w, result, 200)
		case http.MethodPost:
			var body map[string]struct {
				Slots *int `json:"slots"`
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			for svc, cfg := range body {
				state.SetCapacityRule(svc, CapacityRule{Slots: cfg.Slots})
			}
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// GET /_ldk/iam-auth and POST /_ldk/iam-auth
	mux.HandleFunc("/_ldk/iam-auth", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			cfg := state.GetIamConfig()
			mode := "disabled"
			if cfg.Enforce {
				mode = "enforce"
			}
			identitiesOut := make(map[string]interface{})
			for name, identity := range cfg.Identities {
				identitiesOut[name] = identity
			}
			writeJSON(w, map[string]interface{}{
				"mode":              mode,
				"default_identity":  cfg.DefaultIdentity,
				"identities":        identitiesOut,
				"resource_policies": cfg.ResourcePolicies,
			}, 200)
		case http.MethodPost:
			var body map[string]interface{}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			cfg := state.GetIamConfig()

			if mode, ok := body["mode"].(string); ok {
				cfg.Enforce = mode == "enforce"
			}
			if v, ok := body["enforce"].(bool); ok {
				cfg.Enforce = v
			}
			if v, ok := body["default_identity"].(string); ok {
				cfg.DefaultIdentity = v
			}
			if identities, ok := body["identities"].(map[string]interface{}); ok {
				for name, identRaw := range identities {
					identMap, ok := identRaw.(map[string]interface{})
					if !ok {
						continue
					}
					identity := IamIdentity{}
					if policies, ok := identMap["inline_policies"].([]interface{}); ok {
						for _, pRaw := range policies {
							p := parseIamPolicy(pRaw)
							identity.InlinePolicies = append(identity.InlinePolicies, p)
						}
					}
					if bp, ok := identMap["boundary_policy"]; ok && bp != nil {
						p := parseIamPolicy(bp)
						identity.PermissionBoundary = &p
					}
					if bp, ok := identMap["permission_boundary"]; ok && bp != nil {
						p := parseIamPolicy(bp)
						identity.PermissionBoundary = &p
					}
					cfg.Identities[name] = identity
				}
			}
			state.SetIamConfig(cfg)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// POST /_ldk/shutdown
	mux.HandleFunc("/_ldk/shutdown", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.NotFound(w, r)
			return
		}
		writeJSON(w, map[string]string{"status": "shutting down"}, 200)
		if shutdownCh != nil {
			go func() { shutdownCh <- struct{}{} }()
		}
	})

	// GET /_ldk/resources
	mux.HandleFunc("/_ldk/resources", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, map[string]interface{}{"resources": map[string]interface{}{}}, 200)
	})

	// Per-service chaos: PUT/GET/DELETE /_ldk/chaos/{service}
	mux.HandleFunc("/_ldk/chaos/", func(w http.ResponseWriter, r *http.Request) {
		service := strings.TrimPrefix(r.URL.Path, "/_ldk/chaos/")
		if service == "" {
			http.NotFound(w, r)
			return
		}
		switch r.Method {
		case http.MethodPut:
			var body struct {
				ErrorRate float64 `json:"error_rate"`
				LatencyMs int     `json:"latency_ms"`
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			rule := &ChaosRule{
				ErrorRate:    body.ErrorRate,
				LatencyMinMs: body.LatencyMs,
				LatencyMaxMs: body.LatencyMs,
			}
			state.SetChaosRule(service, "*", rule)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		case http.MethodGet:
			svcRules := state.GetAllChaosStatus()
			if info, ok := svcRules[service]; ok {
				writeJSON(w, info, 200)
			} else {
				writeJSON(w, map[string]interface{}{
					"enabled":        false,
					"error_rate":     0,
					"latency_min_ms": 0,
					"latency_max_ms": 0,
				}, 200)
			}
		case http.MethodDelete:
			state.DisableChaos(service)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// Per-service capacity: PUT/GET/DELETE /_ldk/capacity/{service}
	mux.HandleFunc("/_ldk/capacity/", func(w http.ResponseWriter, r *http.Request) {
		service := strings.TrimPrefix(r.URL.Path, "/_ldk/capacity/")
		if service == "" {
			http.NotFound(w, r)
			return
		}
		switch r.Method {
		case http.MethodPut:
			var body struct {
				Slots *int `json:"slots"`
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			state.SetCapacityRule(service, CapacityRule{Slots: body.Slots})
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		case http.MethodGet:
			rule := state.GetCapacityRule(service)
			writeJSON(w, map[string]interface{}{"slots": rule.Slots}, 200)
		case http.MethodDelete:
			state.SetCapacityRule(service, CapacityRule{})
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// Fake server instances: POST/GET /_ldk/fake and GET /_ldk/fake/{name}
	mux.HandleFunc("/_ldk/fake", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodPost:
			var body struct {
				Name     string `json:"name"`
				Endpoint string `json:"endpoint"`
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			state.RegisterFakeServer(body.Name, body.Endpoint)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		case http.MethodGet:
			servers := state.ListFakeServers()
			writeJSON(w, servers, 200)
		default:
			http.NotFound(w, r)
		}
	})

	mux.HandleFunc("/_ldk/fake/", func(w http.ResponseWriter, r *http.Request) {
		name := strings.TrimPrefix(r.URL.Path, "/_ldk/fake/")
		if name == "" {
			http.NotFound(w, r)
			return
		}
		if r.Method != http.MethodGet {
			http.NotFound(w, r)
			return
		}
		endpoint, ok := state.GetFakeServer(name)
		if !ok {
			http.NotFound(w, r)
			return
		}
		writeJSON(w, map[string]string{"name": name, "endpoint": endpoint}, 200)
	})

	// State injection: PUT/GET/DELETE /_ldk/state/{service}/{resourceType}/{resourceId}
	mux.HandleFunc("/_ldk/state/", func(w http.ResponseWriter, r *http.Request) {
		parts := strings.SplitN(strings.TrimPrefix(r.URL.Path, "/_ldk/state/"), "/", 3)
		if len(parts) != 3 {
			writeJSONError(w, "ValidationException", "path must be /_ldk/state/{service}/{resourceType}/{resourceId}", 400)
			return
		}
		service, resourceType, resourceID := parts[0], parts[1], parts[2]
		switch r.Method {
		case http.MethodPut:
			var body struct {
				State string `json:"state"`
			}
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				writeJSONError(w, "ValidationException", "invalid JSON", 400)
				return
			}
			state.SetInjectedState(service, resourceType, resourceID, body.State)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		case http.MethodGet:
			val, ok := state.GetInjectedState(service, resourceType, resourceID)
			if !ok {
				http.NotFound(w, r)
				return
			}
			writeJSON(w, map[string]string{"state": val}, 200)
		case http.MethodDelete:
			state.ClearInjectedState(service, resourceType, resourceID)
			writeJSON(w, map[string]string{"status": "ok"}, 200)
		default:
			http.NotFound(w, r)
		}
	})

	// POST /_ldk/aws-fake — configure fake responses for AWS service operations.
	// Body: { "stepfunctions": { "enabled": true, "rules": [...] } }
	mux.HandleFunc("/_ldk/aws-fake", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.NotFound(w, r)
			return
		}
		var body map[string]map[string]interface{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSONError(w, "ValidationException", "invalid JSON", 400)
			return
		}
		for service, config := range body {
			enabled, _ := config["enabled"].(bool)
			if !enabled {
				state.ClearFakeRules(service)
				continue
			}
			rulesRaw, ok := config["rules"].([]interface{})
			if !ok {
				state.ClearFakeRules(service)
				continue
			}
			var rules []lwsstate.FakeRule
			for _, rRaw := range rulesRaw {
				rMap, ok := rRaw.(map[string]interface{})
				if !ok {
					continue
				}
				operation, _ := rMap["operation"].(string)
				respRaw, _ := rMap["response"].(map[string]interface{})
				var resp lwsstate.FakeResponse
				if respRaw != nil {
					if v, ok := respRaw["status"].(float64); ok {
						resp.Status = int(v)
					}
					if v, ok := respRaw["content_type"].(string); ok {
						resp.ContentType = v
					}
					if v, ok := respRaw["body"].(string); ok {
						resp.Body = v
					}
					if v, ok := respRaw["delay_ms"].(float64); ok {
						resp.DelayMs = int(v)
					}
				}
				rules = append(rules, lwsstate.FakeRule{
					Operation: operation,
					Response:  resp,
				})
			}
			state.SetFakeRules(service, rules)
		}
		writeJSON(w, map[string]string{"status": "ok"}, 200)
	})
}

// wsLogEntry converts a state.LogEntry to the format the SDK's LogCapture expects.
// The SDK's LogEntry uses "handler" for the operation field.
func wsLogEntry(e LogEntry) map[string]interface{} {
	return map[string]interface{}{
		"service":     e.Service,
		"handler":     e.Operation,
		"level":       "info",
		"status_code": e.Status,
		"duration_ms": float64(e.DurationMs),
		"timestamp":   e.Timestamp,
	}
}

func toFloat64(v interface{}) float64 {
	switch t := v.(type) {
	case float64:
		return t
	case float32:
		return float64(t)
	case int:
		return float64(t)
	case int64:
		return float64(t)
	}
	return 0
}

func parseIamPolicy(raw interface{}) IamPolicy {
	policy := IamPolicy{}
	pMap, ok := raw.(map[string]interface{})
	if !ok {
		return policy
	}
	stmtsRaw, ok := pMap["Statement"].([]interface{})
	if !ok {
		return policy
	}
	for _, stmtRaw := range stmtsRaw {
		stmtMap, ok := stmtRaw.(map[string]interface{})
		if !ok {
			continue
		}
		stmt := IamStatement{}
		if v, ok := stmtMap["Effect"].(string); ok {
			stmt.Effect = v
		}
		stmt.Action = stmtMap["Action"]
		stmt.Resource = stmtMap["Resource"]
		policy.Statement = append(policy.Statement, stmt)
	}
	return policy
}

File: lang/typescript/core/src/api/management.ts
/** Management API — /_ldk/* endpoints. */

import type { FastifyInstance, FastifyRequest, FastifyReply } from "fastify";
import type { ServerState, ChaosRule, FakeRule, IamPolicy, CapacityConfig } from "../types";
import { defaultCapacityConfigs } from "../types";
import { WebSocketServer, type WebSocket } from "ws";

// Normalize operation name: "StartExecution" == "start-execution" == "startexecution"
function normalizeOp(op: string): string {
  return op.replace(/[-_]/g, "").toLowerCase();
}

export function registerManagementApi(
  app: FastifyInstance,
  state: ServerState,
  wsServer: WebSocketServer,
): void {
  // GET /_ldk/status
  app.get("/_ldk/status", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({ running: true, providers: [] });
  });

  // POST /_ldk/reset
  app.post("/_ldk/reset", async (_req: FastifyRequest, reply: FastifyReply) => {
    state.chaosRules.clear();
    state.fakeRules.clear();
    state.iamConfig = { enforce: false, identities: {}, resource_policies: {} };
    state.logBuffer = [];

    const defaultConfigs = defaultCapacityConfigs();
    for (const service of Object.keys(state.capacityConfigs)) {
      state.capacityConfigs[service] = { slots: null };
    }
    for (const service of Object.keys(defaultConfigs)) {
      if (!(service in state.capacityConfigs)) {
        state.capacityConfigs[service] = { slots: null };
      }
    }

    for (const cb of state.resetCallbacks) {
      try {
        await Promise.resolve(cb());
      } catch {
        // ignore
      }
    }

    reply.send({ status: "ok" });
  });

  // GET /_ldk/resources
  app.get("/_ldk/resources", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({ resources: {} });
  });

  // GET /_ldk/logs
  app.get("/_ldk/logs", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({ logs: state.logBuffer });
  });

  // GET /_ldk/chaos
  const ALL_SERVICES = [
    "dynamodb",
    "sqs",
    "s3",
    "sns",
    "stepfunctions",
    "events",
    "cognito-idp",
    "ssm",
    "secretsmanager",
  ];
  app.get("/_ldk/chaos", async (_req: FastifyRequest, reply: FastifyReply) => {
    const result: Record<string, unknown> = {};
    for (const service of ALL_SERVICES) {
      const rules = state.chaosRules.get(service);
      const rule = rules?.get("*") ?? {};
      result[service] = {
        enabled: !!rules,
        error_rate: rule.error_rate ?? 0,
        latency_min_ms: rule.latency_min_ms ?? rule.latency_ms ?? 0,
        latency_max_ms: rule.latency_max_ms ?? rule.latency_ms ?? 0,
      };
    }
    reply.send(result);
  });

  // GET /_ldk/iam-auth
  app.get("/_ldk/iam-auth", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({
      mode: state.iamConfig.enforce ? "enforce" : "disabled",
      default_identity: state.iamConfig.default_identity ?? null,
      identities: state.iamConfig.identities,
      resource_policies: state.iamConfig.resource_policies,
    });
  });

  // POST /_ldk/chaos
  // SDK sends: { "service": { "enabled": bool, "error_rate": 0.5, "latency_min_ms": 100, ... } }
  app.post("/_ldk/chaos", async (req: FastifyRequest, reply: FastifyReply) => {
    const body = req.body as Record<string, Record<string, unknown>>;

    for (const [service, config] of Object.entries(body)) {
      const enabled = config.enabled !== false;

      if (!enabled) {
        state.chaosRules.delete(service);
        continue;
      }

      // enabled: true with no other keys = mark service as enabled with empty rules
      if (Object.keys(config).filter((k) => k !== "enabled").length === 0) {
        if (!state.chaosRules.has(service)) {
          state.chaosRules.set(service, new Map());
        }
        continue;
      }

      const errorRate = config.error_rate as number | undefined;
      const latencyMinMs = config.latency_min_ms as number | undefined;
      const latencyMaxMs = config.latency_max_ms as number | undefined;
      const connectionResetRate = config.connection_reset_rate as number | undefined;
      const timeoutRate = config.timeout_rate as number | undefined;
      const errorCode = config.error_code as string | undefined;

      if (!state.chaosRules.has(service)) {
        state.chaosRules.set(service, new Map());
      }
      const serviceRules = state.chaosRules.get(service)!;

      const rule: ChaosRule = {};
      if (errorRate !== undefined && errorRate > 0) rule.error_rate = errorRate;
      if (latencyMinMs !== undefined && latencyMinMs > 0) {
        rule.latency_min_ms = latencyMinMs;
        rule.latency_ms = latencyMinMs;
      }
      if (latencyMaxMs !== undefined && latencyMaxMs > 0) rule.latency_max_ms = latencyMaxMs;
      if (connectionResetRate !== undefined && connectionResetRate > 0) {
        rule.connection_reset = Math.random() < connectionResetRate;
      }
      if (timeoutRate !== undefined && timeoutRate > 0) {
        rule.timeout = Math.random() < timeoutRate;
      }
      if (errorCode) rule.error_code = errorCode;

      // Apply to all operations (*) unless specific operations are configured
      serviceRules.set("*", rule);
    }

    reply.send({ status: "ok" });
  });

  // GET /_ldk/capacity
  app.get("/_ldk/capacity", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send(state.capacityConfigs);
  });

  // POST /_ldk/capacity
  app.post("/_ldk/capacity", async (req: FastifyRequest, reply: FastifyReply) => {
    const body = req.body as Record<string, { slots: number | null }>;
    for (const [service, config] of Object.entries(body)) {
      if (service in state.capacityConfigs) {
        state.capacityConfigs[service] = config as CapacityConfig;
      }
    }
    reply.send({ status: "ok" });
  });

  // POST /_ldk/aws-fake
  // SDK sends: { "service": { "enabled": bool, "rules": [{ "operation": "...", "match_headers": {}, "response": { "status": 200, "content_type": "...", "delay_ms": 0, "body": "..." } }] } }
  app.post("/_ldk/aws-fake", async (req: FastifyRequest, reply: FastifyReply) => {
    const body = req.body as Record<
      string,
      { enabled?: boolean; rules?: Array<Record<string, unknown>> }
    >;

    for (const [service, config] of Object.entries(body)) {
      const enabled = config.enabled !== false;
      const rules = config.rules ?? [];

      if (!enabled || rules.length === 0) {
        state.fakeRules.delete(service);
        continue;
      }

      if (!state.fakeRules.has(service)) {
        state.fakeRules.set(service, new Map());
      }
      const serviceRules = state.fakeRules.get(service)!;

      for (const rule of rules) {
        const operationRaw = rule.operation as string;
        const response = (rule.response ?? {}) as Record<string, unknown>;
        const matchHeaders = (rule.match_headers ?? {}) as Record<string, string>;

        const fakeRule: FakeRule = {
          status: (response.status as number) ?? 200,
          body: response.body,
          delay_ms: (response.delay_ms as number) ?? 0,
          content_type: response.content_type as string | undefined,
          headers: response.content_type
            ? { "Content-Type": response.content_type as string }
            : undefined,
          match_headers: Object.keys(matchHeaders).length > 0 ? matchHeaders : undefined,
        };

        // Check if the body looks like an AWS error
        if (typeof fakeRule.body === "string" && (fakeRule.body as string).includes('"__type"')) {
          try {
            const parsed = JSON.parse(fakeRule.body as string) as Record<string, unknown>;
            if (parsed.__type) {
              fakeRule.error_code = parsed.__type as string;
              fakeRule.error_message = parsed.message as string | undefined;
              fakeRule.http_status = fakeRule.status;
            }
          } catch {
            // keep as-is
          }
        }

        // Store with normalized key (also store by original for exact match)
        serviceRules.set(operationRaw, fakeRule);
        // Also store normalized version
        const normalized = normalizeOp(operationRaw);
        serviceRules.set(`__norm_${normalized}`, fakeRule);
      }
    }

    reply.send({ status: "ok" });
  });

  // POST /_ldk/iam-auth
  // SDK sends: { "mode": "enforce"|"audit"|"disabled", "default_identity": "name", "identities": { "name": { "inline_policies": [...] } } }
  app.post("/_ldk/iam-auth", async (req: FastifyRequest, reply: FastifyReply) => {
    const body = req.body as {
      mode?: string;
      default_identity?: string;
      enforce?: boolean;
      identities?: Record<
        string,
        {
          inline_policies?: IamPolicy[];
          boundary_policy?: IamPolicy;
          permission_boundary?: IamPolicy;
        }
      >;
      resource_policies?: Record<string, IamPolicy>;
    };

    if (body.mode !== undefined) {
      state.iamConfig.enforce = body.mode === "enforce";
      // Store default_identity for lookup
      if (body.default_identity) {
        state.iamConfig.default_identity = body.default_identity;
      }
    }
    if (body.enforce !== undefined) state.iamConfig.enforce = body.enforce;

    if (body.identities) {
      for (const [name, identity] of Object.entries(body.identities)) {
        state.iamConfig.identities[name] = {
          inline_policies: identity.inline_policies,
          permission_boundary: identity.boundary_policy ?? identity.permission_boundary,
        };
      }
    }
    if (body.resource_policies) state.iamConfig.resource_policies = body.resource_policies;

    reply.send({ status: "ok" });
  });

  // POST /_ldk/lifecycle
  app.post("/_ldk/lifecycle", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({ status: "ok" });
  });

  // POST /_ldk/shutdown
  app.post("/_ldk/shutdown", async (_req: FastifyRequest, reply: FastifyReply) => {
    reply.send({ status: "shutting down" });
    setImmediate(() => process.exit(0));
  });

  // WebSocket log streaming at /_ldk/ws/logs
  app.server.on("upgrade", (request, socket, head) => {
    const url = request.url ?? "";
    if (url === "/_ldk/ws/logs") {
      wsServer.handleUpgrade(request, socket, head, (ws: WebSocket) => {
        wsServer.emit("connection", ws, request);
        state.logSubscribers.add(ws);

        // Send buffered logs
        for (const entry of state.logBuffer) {
          try {
            ws.send(JSON.stringify(entry));
          } catch {
            break;
          }
        }

        ws.on("close", () => {
          state.logSubscribers.delete(ws);
        });

        ws.on("error", () => {
          state.logSubscribers.delete(ws);
        });
      });
    }
  });
}

export { normalizeOp };
